### PR TITLE
Task-50284-SignIn SignOut Page

### PR DESCRIPTION
**Pull request by lyxuansang91**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user && user.authenticate(params[:session][:password])
+      sign_in user
+      flash[:success] = "Sign in successfully!"
```
Comment (chuanhtuan) [inline app/controllers/sessions_controller.rb]: i18n

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      redirect_to root_url
+    else
+      flash[:danger] = "invalid email/password. Try again!"
```
Comment (chuanhtuan) [inline app/controllers/sessions_controller.rb]: i18n

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  def show
+  end
+
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      sign_in @user
+      flash[:success] = "Sign up successfully!"
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: i18n

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  def new
+    @user = User.new
+  end
+
+  private
+    def user_params
+      params.required(:user).permit(:name,
+        :email,
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: không xuống dòng kiểu này nhé

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    @current_user = user
+  end
+
+  def current_user
+    remember_token = User.digest cookies[:remember_token]
+    @current_user ||= User.find_by remember_token: remember_token
+  end
+
+  def sign_out
+    current_user.update_attribute :remember_token, User.digest(User.new_remember_token)
```
Comment (chuanhtuan) [inline app/helpers/sessions_helper.rb]: use update_attributes instead of update_attribute

File: app/helpers/sessions_helper.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
+module SessionsHelper
+
+  def sign_in user
+    remember_token = User.new_remember_token
+    cookies.permanent[:remember_token] = remember_token
+    user.update_attribute :remember_token, User.digest(remember_token)
```
Comment (chuanhtuan) [inline app/helpers/sessions_helper.rb]: use update_attributes instead of update_attribute

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
+module SessionsHelper
```
Comment (chuanhtuan) [inline app/helpers/sessions_helper.rb]: anh nghĩ em nên làm giống Tutorial

File: app/views/users/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,22 @@
+<% provide(:title, 'Sign up') %>
+<h1>Sign up</h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+  <%= form_for @user do |f| %>
+    <%= render 'fields', f: f %>
+
+     <!--  <%= f.label :name %>
```
Comment (chuanhtuan) [inline app/views/users/new.html.erb]: remove all remark code

File: app/views/users/_fields.html.erb [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+<%= f.label :name %>
+<%= f.text_field :name %>
+
+<%= f.label :email %>
+<%= f.text_field :email %>
+
+<%= f.label :password %>
+<%= f.password_field :password %>
+
+<%= f.label :password_confirmation, "Confirm Password" %>
```
Comment (chuanhtuan) [inline app/views/users/_fields.html.erb]: i18n

File: config/routes.rb [hunk 10]
```diff
@@ -1,10 +1,19 @@
 Rails.application.routes.draw do
 
+
   root 'static_pages#home'
 
-  match '/help', to: 'static_pages#help',    via: 'get'
-  match '/about', to: 'static_pages#about',   via: 'get'
+  resources :users do
```
Comment (chuanhtuan) [inline config/routes.rb]: không cần do end

File: app/controllers/sessions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,22 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    # byebug
```
Comment (hellovietnam93) [inline app/controllers/sessions_controller.rb]: remove remark

File: app/controllers/sessions_controller.rb [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    if user && user.authenticate(params[:session][:password])
+      sign_in user
+      flash[:success] = t 'login_success'
```
Comment (hellovietnam93) [inline app/controllers/sessions_controller.rb]: "" instead of ''

File: app/controllers/sessions_controller.rb [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+      redirect_to root_url
+    else
+      flash[:danger] = t 'wrong_email_password'
+      render 'new'
```
Comment (hellovietnam93) [inline app/controllers/sessions_controller.rb]: render :new

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -0,0 +1,29 @@
+class UsersController < ApplicationController
+  def index
+  end
+
+  def show
+  end
+
+
+  def create
+    @user = User.new(user_params)
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]:
```
@user = User.new user_params
```

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+    if @user.save
+      sign_in @user
+      flash[:success] = t 'sign_up_success'
+      redirect_back_or root_url
+    else
+      render 'new'
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: render :new

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+  def show
+  end
+
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      sign_in @user
+      flash[:success] = t 'sign_up_success'
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: "" instead of ''

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+    end
+  end
+
+  def new
+    @user = User.new
+  end
+
+  private
+    def user_params
+      params.required(:user).permit(:name, :email, :password,
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]:
```
params.required(:user).permit :name, :email, :password,
 +        :password_confirmation
```

File: app/helpers/sessions_helper.rb [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+  def sign_in user
+    remember_token = User.new_remember_token
+    cookies.permanent[:remember_token] = remember_token
+    user.update_attributes remember_token: User.digest(remember_token)
+    self.current_user = user
+  end
+
+  def current_user=(user)
```
Comment (hellovietnam93) [inline app/helpers/sessions_helper.rb]: bỏ ()

File: app/helpers/sessions_helper.rb [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+
+  def signed_in?
+    !current_user.nil?
+  end
+
+  def current_user? user
+    current_user == user
+  end
+
+  def redirect_back_or(default)
```
Comment (hellovietnam93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+  def store_location
+    session[:return_to] = request.url if request.get?
+  end
+
+  def signed_in_user
+    unless signed_in?
+      store_location
+      redirect_to signin_url, notice: t('please_sign_in')
```
Comment (hellovietnam93) [inline app/helpers/sessions_helper.rb]: "" instead of ''

File: app/helpers/users_helper.rb [hunk 21]
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
```
Comment (hellovietnam93) [inline app/helpers/users_helper.rb]: remove this file if you don't use

File: app/models/user.rb [hunk 22]
```diff
@@ -25,4 +25,23 @@ def User.new_remember_token
...
+  end
+
+
+  def unfollow! other_user
+    relationships.find_by(followed_id: other_user.id).destroy
+  end
+
+  private
+
+    def create_remember_token
```
Comment (hellovietnam93) [inline app/models/user.rb]: ident

File: app/models/user.rb [hunk 23]
```diff
@@ -25,4 +25,23 @@ def User.new_remember_token
...
+    relationships.create!(followed_id: other_user.id)
+  end
+
+
+  def unfollow! other_user
+    relationships.find_by(followed_id: other_user.id).destroy
+  end
+
+  private
+
```
Comment (hellovietnam93) [inline app/models/user.rb]: remove this line

File: app/views/sessions/new.html.erb [hunk 24]
```diff
@@ -0,0 +1,17 @@
+<% provide(:title, t('sign_in')) %>
```
Comment (hellovietnam93) [inline app/views/sessions/new.html.erb]:
```
<% provide :title, t('sign_in') %>
```

File: app/views/sessions/new.html.erb [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+<h1><%= t('sign_in') %></h1>
```
Comment (hellovietnam93) [inline app/views/sessions/new.html.erb]: <h1><%= t "sign_in" %></h1>

File: app/views/shared/_error_messages.html.erb [hunk 26]
```diff
@@ -0,0 +1,12 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger" role="alert">
+      The form contains <%= pluralize(object.errors.count, "error") %>.
```
Comment (hellovietnam93) [inline app/views/shared/_error_messages.html.erb]: I18n

File: app/views/shared/_user_info.html.erb [hunk 27]
```diff
@@ -0,0 +1,7 @@
+<%= image_tag "https://secure.gravatar.com/avatar/bebfcf57d6d8277d806a9ef3385c078d", size: "150x150",  alt: "User Name" %>
+<h2>
+  <%= @current_user.name %>
+</h2>
+<h5 class="text-left">
+  <%= "100 words" %>
```
Comment (hellovietnam93) [inline app/views/shared/_user_info.html.erb]: I18n

File: config/routes.rb [hunk 28]
```diff
@@ -1,10 +1,18 @@
...
+
   root 'static_pages#home'
 
-  match '/help', to: 'static_pages#help',    via: 'get'
-  match '/about', to: 'static_pages#about',   via: 'get'
+  resources :users
+
+  resources :sessions, only: [:new, :create, :destroy]
+
+  match '/help', to: 'static_pages#help', via: 'get'
```
Comment (hellovietnam93) [inline config/routes.rb]: get "help" => "static_pages#help"

File: config/routes.rb [hunk 29]
```diff
@@ -1,10 +1,18 @@
...
+  match '/about', to: 'static_pages#about', via: 'get'
   match '/contact', to: 'static_pages#contact', via: 'get'
+  match '/signup', to: 'users#new', via: 'get'
+  match '/signin', to: 'sessions#new', via: 'get'
+  match '/signout', to: 'sessions#destroy', via: 'delete'
```
Comment (hellovietnam93) [inline config/routes.rb]: delete "logout" => "sessions#destroy"

File: app/helpers/sessions_helper.rb [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+  def signed_in?
+    !current_user.nil?
+  end
+
+  def current_user? user
+    current_user == user
+  end
+
+  def redirect_back_or default
+    redirect_to(session[:return_to] || default)
```
Comment (hellovietnam93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/models/user.rb [hunk 31]
```diff
@@ -15,14 +15,31 @@ class User < ActiveRecord::Base
...
     SecureRandom.urlsafe_base64
   end
 
-
   def User.digest token
     Digest::SHA1.hexdigest token.to_s
   end
+
+   def following? other_user
+    relationships.find_by(followed_id: other_user.id)
```
Comment (hellovietnam93) [inline app/models/user.rb]: remove ()

File: app/models/user.rb [hunk 32]
```diff
@@ -15,14 +15,31 @@ class User < ActiveRecord::Base
...
+    relationships.create!(followed_id: other_user.id)
+  end
+
+  def unfollow! other_user
+    relationships.find_by(followed_id: other_user.id).destroy
+  end
+
+  private
+  def create_remember_token
+      self.remember_token = User.digest User.new_remember_token
```
Comment (hellovietnam93) [inline app/models/user.rb]: ident, can remove ``` self. ``` like: ``` remember_token = User.digest User.new_remember_token ```

File: app/views/layouts/_header.html.erb [hunk 33]
```diff
@@ -4,26 +4,25 @@
       <%#= link_to "sample app", '#', id:"logo", class: "navbar-brand" %>
       <nav>
         <ul class="nav navbar-nav navbar-right">
-          <li><%= link_to "Home", root_url %> </li>
-          <li><%= link_to "Help", help_path %></li>
-          <%# if signed_in? %>
-          <% if false %>
-          <li><%= link_to "Users", "#" %></li>
+          <li><%= link_to t("home"), root_url %> </li>
```
Comment (hellovietnam93) [inline app/views/layouts/_header.html.erb]: remove space before </li>

File: app/views/layouts/_header.html.erb [hunk 34]
```diff
@@ -4,26 +4,25 @@
...
+          <li><%= link_to t("users"), "#" %></li>
           <li class="dropdown" id="dropdownMenuDivider"><a href="#" class="dropdown-toggle" data-toggle="dropdown" aria-expanded="true">
-              Account <b class="caret"></b>
+              <%= t "account" %> <b class="caret"></b>
             </a>
 
             <ul class="dropdown-menu" aria-labelledby="dropdownMenuDivider">
-              <li> <%= link_to "Profile" , "#" %></li>
-               <li><%= link_to "Settings", "#" %></li>
+              <li> <%= link_to t("profile") , @current_user %></li>
```
Comment (hellovietnam93) [inline app/views/layouts/_header.html.erb]: remove space before```,```

File: app/views/layouts/_header.html.erb [hunk 35]
```diff
@@ -4,26 +4,25 @@
...
+               <li><%= link_to t("setting"), "#" %></li>
               <!-- <li class="disabled"><a href="#">Disabled link</a></li> -->
               <li class="divider"></li>
-              <!-- <li> <% link_to("Sign Out", "#" , method: "delete") %></li> -->
+             <li> <%= link_to t("sign_out"), signout_path , method: "delete" %></li>
```
Comment (hellovietnam93) [inline app/views/layouts/_header.html.erb]: remove space

File: app/views/sessions/new.html.erb [hunk 36]
```diff
@@ -0,0 +1,17 @@
+<% provide :title, t("sign_in") %>
+<h1><%= t "sign_in"  %></h1>
```
Comment (hellovietnam93) [inline app/views/sessions/new.html.erb]: remove one space before %>

File: app/views/shared/_activity.html.erb [hunk 37]
```diff
@@ -0,0 +1,23 @@
+<div class="activities">
+<div class="row">
+  <%= image_tag "https://secure.gravatar.com/avatar/bebfcf57d6d8277d806a9ef3385c078d",  size: "50x50",  alt: "User Name" %>
```
Comment (hellovietnam93) [inline app/views/shared/_activity.html.erb]: this line is too long, divide into 2 lines

File: app/views/shared/_error_messages.html.erb [hunk 38]
```diff
@@ -0,0 +1,14 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger" role="alert">
+
+    <%= t "form_contain" %>
+       <%= pluralize(object.errors.count, "error") %>.
```
Comment (hellovietnam93) [inline app/views/shared/_error_messages.html.erb]: <%= pluralize object.errors.count, "error" %>.

Comment (hellovietnam93) [inline app/views/shared/_error_messages.html.erb]: ident

File: app/views/shared/_error_messages.html.erb [hunk 39]
```diff
@@ -0,0 +1,14 @@
...
+    </div>
+    <ul>
+    <% object.errors.full_messages.each do |msg| %>
```
Comment (hellovietnam93) [inline app/views/shared/_error_messages.html.erb]: ident

File: app/views/shared/_user_info.html.erb [hunk 40]
```diff
@@ -0,0 +1,7 @@
+<%= image_tag "https://secure.gravatar.com/avatar/bebfcf57d6d8277d806a9ef3385c078d", size: "150x150",  alt: "User Name" %>
```
Comment (hellovietnam93) [inline app/views/shared/_user_info.html.erb]: divide into 2 lines

File: app/views/static_pages/home.html.erb [hunk 41]
```diff
@@ -1,32 +1,31 @@
...
   <div class="col-md-4 border-right">
     <%= render 'shared/user_info' %>
   </div>
 
-
   <div class="col-md-8">
     <div class="row">
       <div class="col-md-6">
-      <%= link_to "Words", words_path, class: "btn btn-default btn-primary full-size" %>
+      <%= link_to t('words'), "#", class: "btn btn-default btn-primary full-size" %>
```
Comment (hellovietnam93) [inline app/views/static_pages/home.html.erb]: using ""

File: app/views/static_pages/home.html.erb [hunk 42]
```diff
@@ -1,32 +1,31 @@
...
       </div>
 
       <div class="col-md-6">
-      <%= link_to "Lessons", categories_path, class: "btn btn-default btn-primary full-size"%>
+      <%= link_to t('lessons'), "#", class: "btn btn-default btn-primary full-size"%>
       </div>
     </div>
-  <h3>User Activities</h3>
+  <h3><%= t('user_activities')  %></h3>
```
Comment (hellovietnam93) [inline app/views/static_pages/home.html.erb]: remove ()

File: app/views/users/_fields.html.erb [hunk 43]
```diff
@@ -0,0 +1,13 @@
+<%= render 'shared/error_messages', object: f.object %>
```
Comment (hellovietnam93) [inline app/views/users/_fields.html.erb]: using ""

File: config/routes.rb [hunk 44]
```diff
@@ -1,10 +1,18 @@
...
+
+  resources :users
+
+  resources :sessions, only: [:new, :create, :destroy]
+
+  get "/help" => "static_pages#help"
+  get "/about" => "static_pages#about"
+  get "/contact" => "static_pages#contact"
+  get "/signup" => "users#new"
+  get "/signin" => "sessions#new"
```
Comment (hellovietnam93) [inline config/routes.rb]: lack of method post

File: app/views/layouts/_header.html.erb [hunk 45]
```diff
@@ -4,26 +4,25 @@
...
 
             <ul class="dropdown-menu" aria-labelledby="dropdownMenuDivider">
-              <li> <%= link_to "Profile" , "#" %></li>
-               <li><%= link_to "Settings", "#" %></li>
+              <li> <%= link_to t("profile"), @current_user %></li>
+               <li><%= link_to t("setting"), "#" %></li>
               <!-- <li class="disabled"><a href="#">Disabled link</a></li> -->
               <li class="divider"></li>
-              <!-- <li> <% link_to("Sign Out", "#" , method: "delete") %></li> -->
+             <li> <%= link_to t("sign_out"), signout_path, method: "delete" %></li>
```
Comment (chuanhtuan) [inline app/views/layouts/_header.html.erb]: method: :delete

File: app/views/layouts/_header.html.erb [hunk 46]
```diff
@@ -4,26 +4,23 @@
...
         <ul class="nav navbar-nav navbar-right">
-          <li><%= link_to "Home", root_url %> </li>
-          <li><%= link_to "Help", help_path %></li>
-          <%# if signed_in? %>
-          <% if false %>
-          <li><%= link_to "Users", "#" %></li>
+          <li><%= link_to t("home"), root_url %></li>
+          <li><%= link_to t("help"), help_path %></li>
+          <% if signed_in? %>
+          <li><%= link_to t("users"), "#" %></li>
```
Comment (hellovietnam93) [inline app/views/layouts/_header.html.erb]: ident

File: app/views/static_pages/home.html.erb [hunk 47]
```diff
@@ -1,32 +1,31 @@
-<% if 1 < 0 %>
+<% if signed_in? %>
 <div class="row">
```
Comment (hellovietnam93) [inline app/views/static_pages/home.html.erb]: ident

File: app/views/users/new.html.erb [hunk 48]
```diff
@@ -0,0 +1,10 @@
+<% provide :title, t("sign_up") %>
+<h1><%= t "sign_up" %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+  <%= form_for @user do |f| %>
+    <%= render 'fields', f: f %>
```
Comment (hellovietnam93) [inline app/views/users/new.html.erb]: using ""

File: app/views/static_pages/home.html.erb [hunk 49]
```diff
@@ -1,34 +1,24 @@
...
-      <%= link_to "Lessons", categories_path, class: "btn btn-default btn-primary full-size"%>
+<% if signed_in? %>
+  <div class="row">
+    <div class="col-md-4 border-right">
+      <%= render "shared/user_info" %>
+    </div>
+    <div class="col-md-8">
+      <div class="row">
+        <div class="col-md-6">
+        <%= link_to t("words"), "#", class: "btn btn-default btn-primary full-size" %>
```
Comment (hellovietnam93) [inline app/views/static_pages/home.html.erb]: ident

File: app/views/static_pages/home.html.erb [hunk 50]
```diff
@@ -1,34 +1,24 @@
...
+        </div>
+        <div class="col-md-6">
+        <%= link_to t("lessons"), "#", class: "btn btn-default btn-primary full-size"%>
```
Comment (hellovietnam93) [inline app/views/static_pages/home.html.erb]: ident

File: app/helpers/sessions_helper.rb [hunk 51]
```diff
@@ -0,0 +1,49 @@
...
+    session[:return_to] = request.url if request.get?
+  end
+
+  def signed_in_user
+    unless signed_in?
+      store_location
+      redirect_to signin_url, notice: t("please_sign_in")
+    end
+  end
+
```
Comment (hellovietnam93) [inline app/helpers/sessions_helper.rb]: remove this line

File: app/models/user.rb [hunk 52]
```diff
@@ -15,14 +15,31 @@ class User < ActiveRecord::Base
...
+
+  def unfollow! other_user
+    relationships.find_by(followed_id: other_user.id).destroy
+  end
+
+  private
+  def create_remember_token
+    remember_token = User.digest User.new_remember_token
+  end
+
```
Comment (hellovietnam93) [inline app/models/user.rb]: remove this line

File: app/views/shared/_error_messages.html.erb [hunk 53]
```diff
@@ -0,0 +1,13 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger" role="alert">
+    <%= t "form_contain" %>
+       <%= pluralize object.errors.count, "error" %>.
```
Comment (hellovietnam93) [inline app/views/shared/_error_messages.html.erb]: ident, 1 tab

File: app/views/static_pages/help.html.erb [hunk 54]
```diff
@@ -1,5 +1,5 @@
-<% provide(:title, 'Help') %>
-<h1>Help</h1>
+<% provide :title, t("help") %>
+<h1><%= t "help" %></h1>
 <p>
-<h5>This is help page</h5>
+<h5><%= t "help_page" %></h5>
```
Comment (hellovietnam93) [inline app/views/static_pages/help.html.erb]: ident

File: app/views/layouts/application.html.erb [hunk 55]
```diff
@@ -1,26 +1,23 @@
 <!DOCTYPE html>
 <html>
 <head>
-  <title><%= full_title(yield(:title)) %></title>
-  <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track' => true %>
-  <%= javascript_include_tag 'application', 'data-turbolinks-track' => true %>
+  <title><%= full_title yield(:title) %></title>
+  <%= stylesheet_link_tag "application", media: "all", "data-turbolinks-track" => true %>
```
Comment (hellovietnam93) [inline app/views/layouts/application.html.erb]: ở các bản ruby mới thì sử dụng như thế này được nhé,
```<%= stylesheet_link_tag "application", media: "all", "data-turbolinks-track": true %>```

File: app/views/users/new.html.erb [hunk 56]
```diff
@@ -0,0 +1,10 @@
+<% provide :title, t("sign_up") %>
+<h1><%= t "sign_up" %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+  <%= form_for @user do |f| %>
+    <%= render "fields", f: f %>
+      <%= f.submit t("create_account"), class: "btn btn-large btn-primary" %>
```
Comment (hellovietnam93) [inline app/views/users/new.html.erb]: ident

**Comment by hellovietnam93** [comment]
Ok

**Pull request merged**